### ISSUE TITLE
Command naming consistency

### ISSUE THREAD

**Issue by 0x2b3bfa0** [iterative/cml#762]
Unify naming [without breaking backwards compatibility](https://github.com/yargs/yargs/blob/main/docs/advanced.md#command-aliases).

- more intuitive (e.g. `cml <noun> <verb> [options]`)
- easier to explain
- better correspondence with (to be) documented use cases

| Current | Proposed |
| --- | --- |
| `cml ci` | `cml ci [setup]` |
| `cml pr` | `cml pr [create] [{--merge,--rebase,--squash}]` | 
| | `cml pr {create <file...>,close [<number>]}` |
| | [`cml pr {merge} [{--no-ff(?),--ff(?),--rebase,--squash}]`](https://github.com/iterative/cml/issues/823#issuecomment-982348333) |
| `cml publish [--native] [--type=file]` | *hide* |
| `cml runner` | `cml runner [launch]` |
| `cml send-comment` | **`cml {comment,report} {create,update}`** |
| | [`cml {comment,report} watch [--update]`](https://github.com/iterative/cml/pull/1026#discussion_r901479836) |
| `cml send-github-check` | **`cml check {create,update}`** |
| `cml tensorboard-dev` | `cml tensorboard-dev [publish]` |
| `cml rerun-workflow` | **[`cml workflow {rerun,stop}`](https://github.com/iterative/cml/issues/827#issuecomment-985447709)** |

### See also

* https://primer.style/cli/foundations/language
* https://github.com/spf13/cobra#concepts

### Related
* https://github.com/iterative/cml/issues/823#issuecomment-982348333
* [x] https://github.com/iterative/cml/pull/1026#discussion_r898742283  — unify trailing periods on `--help` messages

**Comment by DavidGOrtega** [comment]
- pr generates a pr
 - comment generates a comment
 - check generates a check
 - runner generates a runner
 - tensorboard-dev generates a resource at tb dev
 
then we have ```publish``` its in reality the outsider... ```media```?

**Comment by casperdcl** [comment]
1. Updated description to be `cml noun {--mutex-verb}`. With aliases, this retains backwards-compatibility
2. I'd also say default-verbs don't need be implemented (e.g. `cml pr` without flags obviously implies *create*, so we don't need to support an explicit `--create` flag).

**Comment by 0x2b3bfa0** [comment]
* https://github.com/iterative/cml/issues/823#issuecomment-982348333

**Comment by casperdcl** [comment]
TL;DR for now we should do:

- send-comment -> {comment,report} [{--create,--update}]
- rerun-workflow -> workflow [--rerun]

**Comment by 0x2b3bfa0** [comment]
If we're supporting actions other than `create` commands should be in the form `cml noun verb --modifier` instead of `cml noun --verb --modifier` as @casperdcl proposed.

If there isn't any command supporting verbs other than `create` we may make it implicit as well, and have `cml noun --modifier`

**Comment by dacbd** [comment]
`cml noun verb[implied create] --modifier` +1

**Comment by casperdcl** [comment]
Hmm. `cml report [create]` won't work. `create` will have to be `<required>` (not `[optional]`) due to the next arg also being a `<positional>`: `cml comment {create,update} <report.md>`.

Isn't `cml comment [{--create,--update}] <report.md>` nicer?

**Comment by 0x2b3bfa0** [comment]
> _Isn't `cml comment [{--create,--update}] <report.md>` nicer?_

🐻👨🏼‍🍳🤺 [No strong opinions on this](https://youtu.be/T0eEBsWOSVU?t=3543) but things get a bit messy when using `--options` both for verbs and modifiers. E.g. `cml report --update --pr --commit-sha=2b3bfa0` 🤔

**Comment by 0x2b3bfa0** [comment]
Anyway, my biggest concern is extensibility. For the current command set, it doesn't look that bad.

**Comment by 0x2b3bfa0** [comment]
Request for ~last will~ comments

| Current | Proposed |
| --- | --- |
| `cml ci` | **`cml ci setup`** |
| `cml pr` | `cml pr [create] <file...>` | 
| | `cml pr close [number]` |
| | [`cml pr merge [--rebase\|--squash]`](https://github.com/iterative/cml/issues/823#issuecomment-982348333) |
| `cml publish` | _hidden, superseded by `cml report`[^1]_ |
| `cml rerun-workflow` | **[`cml workflow <restart\|stop>`](https://github.com/iterative/cml/issues/827#issuecomment-985447709)** |
| `cml runner` | **`cml runner start`**[^2] |
| `cml send-comment` | **`cml report <create\|update>`** |
| `cml send-github-check` | **`cml check <create\|update>`** |
| `cml tensorboard-dev` | **`cml tensorboard start`** |

[^1]: Since there are no external uses for `cml publish` beyond `cml report` and `cml check` and #1026 eliminates the need for a separate command
[^2]: While register/unregister is semantically equivalent to create/delete, the unregister part is not exposed to users, and **start** feels more natural; i.e. same fire-and-forget behavior as `cml tensorboard start`

**Comment by dacbd** [comment]
`cml runner [start]` ?

**Comment by casperdcl** [comment]
Surely `cml pr [create]` -> `cml pr create`?

**Comment by casperdcl** [comment]
bumping priority on this (blocking everythign from https://github.com/iterative/cml/pull/1026#discussion_r901479836 to https://github.com/iterative/cml.dev/issues/250)

**Comment by casperdcl** [comment]
I see 3 points of contention, can vote with emojis:

1. cml runner
  a) [launch] :+1: 
  b) [start] :eyes:
2. cml tensorboard-dev
  a) [publish] :smile:
  b) [start] :tada:
3. cml workflow
  a) rerun :heart:
  b) restart :rocket:

**Comment by 0x2b3bfa0** [comment]
_[Philosophical question (bis): are checks and comments two different kinds of report?](https://github.com/iterative/cml/issues/762#issuecomment-1156831989)_

**Comment by 0x2b3bfa0** [comment]
🛎️ @iterative/cml: _[philosophical question (bis): are checks and comments two different kinds of report?](https://github.com/iterative/cml/issues/762#issuecomment-1156831989)_

**Comment by 0x2b3bfa0** [comment]
How is `cml <report|check> create` different from `cml <report|check> update` in behavior? Doesn't the latter imply the former?[^1] Would it make more sense to provide `cml <report|check> create --update` instead?

[^1]: i.e. for practical reasons, `update` means _create if not exists, update otherwise_

**Comment by casperdcl** [comment]
- btw https://cli.github.com/manual/gh_run_rerun & https://cli.github.com/manual/gh_workflow_run strongly implies (3a:heart:) beats (3b:rocket:)
- `report create` vs `report update` - yes, they are different. "Create" always makes a new comment, "update" tries to edit an existing comment
- `report create --update` seems as illogical to me as `s/workflow rerun/workflow run --rerun/`

**Comment by 0x2b3bfa0** [comment]
> _"update" tries to edit an existing comment_

And what if there isn't an existing comment to edit?

**Comment by DavidGOrtega** [comment]
> How is cml <report|check> create different from cml <report|check> update in behavior? Doesn't the latter imply the former?[1](https://github.com/iterative/cml/issues/762#user-content-fn-1-4adf7a71dfd8d13b0517603b37f4147d) Would it make more sense to provide cml <report|check> create --update instead?

`check` should be able to offer the capability of guessing the status with code checks, like deepchecks or easy comparator s over dvc diff metrics...

**Comment by 0x2b3bfa0** [comment]
_Follow-up https://github.com/iterative/cml/pull/1026#discussion_r901479836 probably required._

**Comment by casperdcl** [comment]
Yes, exactly. Strong preference for

- consistency (`cml <noun> <verb> [options]`)
- conciseness ("`update`: updates existing CML comment, falling back to `create` if none exists" is fine to document IMO, while `createorupdate` is clunky and requires precisely the same docstring anyway.)

So  means it must be `cml report {create,read,update,delete}`...

**Comment by 0x2b3bfa0** [comment]
> _`createorupdate` is clunky_

Clarification: I never meant to have a verb named `createorupdate`; I used that string only for the sake of clarity on this conversation.

**Comment by casperdcl** [comment]
Not to me. Seems same level as "create" (vis. the CRUD acronym)
